src/test/run-make/sanitizer-staticlib-link/Makefile

```diff
@@ -0,0 +1,25 @@
+-include ../tools.mk
+
+# This test builds a staticlib, then an executable that links to it.
+# The staticlib and executable both  are compiled with address sanitizer, 
+# and we assert that a fault in the staticlib is correctly detected.
+
+# NOTE the address sanitizer only supports x86_64 linux and macOS
+
+ifeq ($(TARGET),x86_64-apple-darwin)
+ASAN_SUPPORT=$(SANITIZER_SUPPORT)
+EXTRA_RUSTFLAG=-C rpath
+else
+ifeq ($(TARGET),x86_64-unknown-linux-gnu)
+ASAN_SUPPORT=$(SANITIZER_SUPPORT)
+EXTRA_RUSTFLAG=
+endif
+endif
+
+all:
+ifeq ($(ASAN_SUPPORT),1)
+	$(RUSTC) -g -Z sanitizer=address --crate-type staticlib --target $(TARGET) library.rs
+	$(CC) program.c $(call STATICLIB,library) $(call OUT_EXE,program) -lasan $(EXTRACFLAGS) $(EXTRACXXFLAGS)
```

Comment:
It looks like when I move staticlib to "linking" the runtime in creader.rs, the runtime is actually not linked. When I do readelf on the static lib I see:

```
     2: 0000000000000000    28 OBJECT  LOCAL  DEFAULT   17 .L__asan_gen_
     3: 000000000000001c    28 OBJECT  LOCAL  DEFAULT   17 .L__asan_gen_.1
     4: 0000000000000038    13 OBJECT  LOCAL  DEFAULT   17 .L__asan_gen_.2
    10: 0000000000000000    32 FUNC    LOCAL  DEFAULT   10 asan.module_ctor
    11: 0000000000000000    22 FUNC    LOCAL  DEFAULT   12 asan.module_dtor
    28: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_init
    29: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_option_detect_stack_use_after_return
    30: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_register_globals
    31: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_report_load4
    32: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_report_load8
    33: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_report_store4
    34: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_report_store8
    35: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_stack_malloc_0
    36: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_stack_malloc_1
    37: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_unregister_globals
    38: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __asan_version_mismatch_check_v8
```

This is why I tried to use -lasan without the runtime link setting in creader. This makes me think that somewhere when linking the staticlib type we aren't actually linking the asan runtime despite setting this to true. 

An option is to remove the staticlib support for now and to continue with out, but I would like to try and get it to operate.